### ISSUE TITLE
RFE: Block Acceleration/Gyroscope Tracking

### ISSUE THREAD

**Issue by TPS** [AdguardTeam/AdguardForAndroid#1436]
___N.B.:  I'm really @ a loss as to _where_ to place this RFE (or if AG can/should do anything @ all about it).  Also, the amount of accessible different mechanisms available to track anything & everything 1 does grows increasingly ridiculous! 😤___

It's now possible to track what 1 does in the physical realm using simple JavaScript & an overly permissive engine.

### Steps to reproduce

POC: https://krausefx.github.io/whats-the-user-doing/
Repo: https://github.com/KrauseFx/whats-the-user-doing/
Code: https://github.com/KrauseFx/whats-the-user-doing/blob/master/index.html

### Expected behavior

In theory, this should require a separate permission to do, or @ least a prompt before using the gyro sensors.

### Actual behavior

No prompt nor block. 😞

### Your environment

* Adguard version: 2.10.56 Amazon ß
* Adguard filtering mode: VPN
* Device model: (e.g. Google Pixel) Nextbit → Razer Robin
* Operating system and version: Android 7.1.1
* Rooted or not? Not
* Any other apps that can start VPN/Antiviruses/Cleaners etc. installed? N/A

**Comment by TPS** [comment]
P.S.: It's now a JS library @ https://github.com/cagataycali/whats-the-user-doing.js for complete script-kiddy convenience. 😶

**Comment by ameshkov** [comment]
This is interesting indeed, and it should ask for a permission -- I don't think that blocking it entirely is a good solution.

Let's conduct an experiment.

Add this rule and check how many websites are really trying to exploit it:
```
#%#(function(){var b=null,e=["devicemotion","deviceorientation"],c=window.addEventListener;window.addEventListener=function(a,d){-1===e.indexOf(a)?c(a,d):(null===b&&(b=confirm("Page tries to access your device's gyro data. Do you grant the permission?")),b?c(a,d):console.log("Denied access to "+a))}})();
```

For instance, it will ask for your permission on that test page.

It's important to note, that for it to function AG needs to be able to modify pages content, therefore HTTPS filtering is mandatory (only in your browser, though).

**Comment by TPS** [comment]
What platform/browser are you using?  I'm not getting _any_ permission prompts @ _all_!

**Comment by ameshkov** [comment]
Let me check on Android:)

**Comment by TPS** [comment]
Oh, wait....  Add that function where?  AG user filters?  I misread what you meant.

**Comment by ameshkov** [comment]
> AG user filters?

Yeah, just add this exact rule to the AG user filter

**Comment by ameshkov** [comment]
Works for me in Chrome

**Comment by TPS** [comment]
Rebooting in attempt to activate HTTPS filtering to try it out....  Gimme a few moments.

**Comment by TPS** [comment]
Nope, nothing, but I'm only using WebView browsers, most LastPass.

**Comment by ameshkov** [comment]
HTTPS filtering is disabled for LastPass by default

**Comment by TPS** [comment]
Doesn't work in Firefox Klar (F-droid), either. Is _that_ whitelisted?

**Comment by ameshkov** [comment]
It targets Nougat, hence it does not trust user certificates: https://blog.adguard.com/en/android-nougat-release-and-what-does-it-mean-for-adguard-users/

Most major browsers made an exception for it and opted to trust user certificates (Chrome, Samsung, regular FF if you install the certificate into their cert storage).

However, there're some browsers that don't do it, like Firefox Focus for instance.

**Comment by ameshkov** [comment]
Here is the GH issue where we test it: #1304

**Comment by TPS** [comment]
I'm _running_ Nougat, so does that mean I'm SOL re: HTTPS-filtering?  I'd be ok w/ that, anyway, but just to try this POC-fix, I've enabled it.

**Comment by ameshkov** [comment]
> SOL re: HTTPS-filtering

Mm? Can't parse it:)

**Comment by TPS** [comment]
SOL = S*** Outta Luck....  Sorry. =)   I mean, Nougat _can't_ HTTPS-filter, then?

**Comment by ameshkov** [comment]
> SOL = S*** Outta Luck.... Sorry. =) I mean, Nougat can't HTTPS-filter, then?

It can, but the app dev should allow it by adding special "network security configuration".

Here is a list of major browsers supporting it:
```
Chrome
Samsung
Firefox (you'll need to tap "install certificate into Firefox" in AG settings)
Opera
Dolphin
Yandex
```

**Comment by TPS** [comment]
Meh, I'll just turn HTTPS-filtering off.  But if the POC-fix works, maybe include that in the experimental filters & find some beta testers to try it on mobile?

**Comment by ameshkov** [comment]
I have it in the user filter myself, we'll see if anybody tries to exploit it.

**Comment by ameshkov** [comment]
Here's how it looks like:
https://uploads.adguard.com/up04_2xrsf.png

**Comment by TPS** [comment]
Does it actually grant/deny the permission?  The JS seems to just log the decision.

**Comment by ameshkov** [comment]
It does

**Comment by TPS** [comment]
Here's a non-HTTPS site for which I get the function prompt:

http://www.islandpacket.com/opinion/opn-columns-blogs/david-lauderdale/article135539793.html

How do I confirm the block?  I don't even see anything in the log: 
[adguard_logs_1709_2248.zip](https://github.com/AdguardTeam/AdguardForAndroid/files/1309597/adguard_logs_1709_2248.zip)

**Comment by TPS** [comment]
Here's another: http://m.starwoodhotels.com/westin/property/area/transportation.html?propertyID=1415&language=en_US

**Comment by ameshkov** [comment]
`islandpacket.com` uses it for legit purposes (changes the page layout on device orientation change).

`m.starwoodhotels.com` -- not so clear here, the script is heavily obfuscated, I can't understand why does it need it yet.

**Comment by TPS** [comment]
Changing page layout doesn't need gyro - 1 can get screen/window dimensions to do _that_.  But blocking them makes me happier, anyway.

**Comment by ameshkov** [comment]
It detects the portrait<->album transitions, nothing serious really.

**Comment by TPS** [comment]
_Just FYI & OT: If ever translating the change from & to portrait orientation, the other 1 is called_ landscape _in the US._

**Comment by TPS** [comment]
Folks @ Princeton U. have juiced this up _(like on **steroids**!)_ & made a patent-pending version, it seems:

https://www.princeton.edu/news/2017/11/29/phones-vulnerable-location-tracking-even-when-gps-services

Whaddy'all think, @AdGuardTeam?

**Comment by ameshkov** [comment]
Not that we can do anything with that :(

**Comment by TPS** [comment]
Another example for the original issue:  Any page @

http://www.star-telegram.com/

**Comment by TPS** [comment]
_NB: I recommend doing this link on a private browser/tab, or 1 that all tracking data can be wiped, as article accesses are limited monthly._

Yet another fine example @ https://www.bloomberg.com/news/features/2018-09-20/making-marines-into-macgyvers

**Comment by TPS** [comment]
@ameshkov A further idea for this:  Maybe have a toggle/explanation for this in the Stealth module/tab, which just [adds your rule above](https://github.com/AdguardTeam/AdguardForAndroid/issues/1436#issuecomment-328379256), since that still works fine.  Y'all could even default it off.

**Comment by artemiv4nov** [comment]
It looks like this request is outdated. Given that only a few people voted, I’m closing the task

**Comment by TPS** [comment]
Just found out Google itself has implemented part of this, e.g., https://www.google.com/search?q=bubble+level as part of some builtin tools.
